File: synapses/py/vision_messages.py
"""
UDP message contracts shared with synapses/ts. Keep in sync when adding fields.
"""

import json
from dataclasses import dataclass
from typing import Any

# --- UDP 5005: Nervous System → Eyes ---

@dataclass
class LookCommand:
    action: str = "look"
    x: float = 0.0
    y: float = 0.0
    pupil_mode: str = "normal"

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> "LookCommand":
        return cls(
            action=data.get("action", "look"),
            x=float(data.get("x", 0)),
            y=float(data.get("y", 0)),
            pupil_mode=str(data.get("pupil_mode", "normal")),
        )

    def to_json(self) -> str:
        out = {"action": self.action, "x": self.x, "y": self.y}
        if self.pupil_mode != "normal":
            out["pupil_mode"] = self.pupil_mode
        return json.dumps(out)


@dataclass
class EyeTrackingEvent:
    """UDP 5006: Camera / eye-track → Nervous System."""

    event: str
    label: str | None = None
    confidence: float | None = None
    x: float | None = None
    y: float | None = None

    def to_json(self) -> str:
        d: dict[str, Any] = {"event": self.event}
        if self.label is not None:
            d["label"] = self.label
        if self.confidence is not None:
            d["confidence"] = self.confidence
        if self.x is not None:
            d["x"] = self.x
        if self.y is not None:
            d["y"] = self.y
        return json.dumps(d)

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> "EyeTrackingEvent":
        return cls(
            event=str(data.get("event", "")),
            label=data.get("label"),
            confidence=data.get("confidence"),
            x=data.get("x"),
            y=data.get("y"),
        )

File: synapses/ts/vision_messages.ts
/**
 * UDP 5005: Nervous System → Eyes
 * All commands are JSON with an "action" field; main_eyes.py dispatches by action.
 */

export interface LookCommand {
	action: "look";
	x: number; // -1.0 to 1.0
	y: number; // -1.0 to 1.0
	pupil_mode?: "normal" | "wide" | "narrow";
}

export interface AnimationCommand {
	action: "animation";
	name: string; // e.g. "double_blink" | "nervous_look" | "shiver" | "sleep_close"
	replace?: boolean;
}

export interface EyesOpenCommand {
	action: "eyes_open";
}

export interface EyesCloseCommand {
	action: "eyes_close";
}

export interface SleepCloseCommand {
	action: "sleep_close";
	duration_s?: number;
}

export interface WarmupCommand {
	action: "warmup";
	step: number;
}

export interface SetEyeShapeCommand {
	action: "set_eye_shape";
	shape: string;
}

export interface SetEyeTypeCommand {
	action: "set_eye_type";
	type?: string;
	eye_type?: string;
}

export interface CycleEyeTypeCommand {
	action: "cycle_eye_type";
}

export interface BlinkCommand {
	action: "blink";
}

export interface ImpulseCommand {
	action: "impulse";
}

export interface RestartBothCommand {
	action: "restart_both";
}

export type EyesCommand =
	| LookCommand
	| AnimationCommand
	| EyesOpenCommand
	| EyesCloseCommand
	| SleepCloseCommand
	| WarmupCommand
	| SetEyeShapeCommand
	| SetEyeTypeCommand
	| CycleEyeTypeCommand
	| BlinkCommand
	| ImpulseCommand
	| RestartBothCommand;

/**
 * UDP 5006: Camera / eye-track → Nervous System
 */
export interface EyeTrackingEvent {
	event:
		| "eye_tracking_started"
		| "eye_tracking_stopped"
		| "looking_started"
		| "looking_stopped"
		| "looking_at";
	label?: string;
	confidence?: number;
	x?: number; // raw pixel or normalized
	y?: number;
}
